content/blog/2019-10-02-codestitcher.md

```diff
@@ -0,0 +1,166 @@
+
+Another drawback of PH is that it uses an undirected call graph. As a result, in the eyes of PH, function *A* calling *B* is the same as *B* calling *A*, and PH treats ordering *A* before *B* as the same as ordering *B* before *A*. However, this is clearly not the case. Let's say *A* calls *B* 100 times, and *B* calls *A* 0 times. Then, if we order *A* before *B*, assuming the call occurs in the middle of *A*, the instruction pointer will have to jump *len(A)/2* instructions to the top of *B*, where *len(A)* denotes the number of instructions in *A*. Alternatively, if *B* was placed before *A*, then *A* would on average jump *len(B) + len(A)/2* instructions. Because of this discrepancy, it is important to take into account the caller and callee relations in the call graph.
+
+### Call-chain clustering
+[Call-chain clustering (C3)](https://research.fb.com/publications/optimizing-function-placement-for-large-scale-data-center-applications-2/) is a recent function layout optimization developed at Facebook Research that addresses the issue with using undirected call graphs. We do not discuss its internal details in this post, but a simplified way of thinking about C3 is that it is PH but with a directed call graph. In this way, it improves spatial locality by considering caller-callee relations.
+
+## Codestitcher
+
+This brings us to Codestitcher, an optimizer that integrates basic block reordering and function reordering to improve upon the PH model. Below is a high-level overview of the steps taken in Codestitcher to generate the optimized code layout, given an input program.
+
+1. Compile the program to LLVM bitcode. This serves as the baseline program. As the program is built, symbols are emitted that allow us to generate a CFG for the program.
+2. Profile the program, generating weights for CFG edges.
+3. Using the weighted CFG, perform basic block chaining, ordering basic blocks in a way to minimize jumps.
+4. Perform hierarchical code collocation, ordering these basic block chains to form higher-level layouts. Here, the *d*-close partial layout algorithm works to build layouts to maximize spatial locality.
+
+There is not much to say about step 1, so let us explore steps 2, 3, and 4 in more detail.
+
+### Profiling
+In order to build our weighted CFG for layout analysis, we need to collect data about "control flow transfers" (when control transfers between CFG nodes). Luckily, most Intel CPUs come with a "last branch record", or LBR--- storing recently executed instructions such as jumps, branches, and calls. With the Linux `perf` tool, this data can be gathered and used to generate a weighted CFG for optimizing program layout.
```

Comment:
```suggestion
In order to build our weighted CFG for layout analysis, we need to collect data about "control flow transfers" (when control transfers between CFG nodes). Luckily, most Intel CPUs come with a "last branch record", or LBR---storing recently executed instructions such as jumps, branches, and calls. With the Linux `perf` tool, this data can be gathered and used to generate a weighted CFG for optimizing program layout.
```

Reply:
While nifty, I agree that this only deserves as much space as you give it. This discussion can probably be folded into the overview just to describe where the weighted CFG comes from.